app/models/fire_cloud_client.rb

```diff
@@ -1332,6 +1336,61 @@ def delete_user_from_billing_project(project_id, role, email)
     end
   end
 
+  ##
+  # PET SERVICE ACCOUNT METHODS
+  # these methods reference the SAM API directly for issuing pet service account tokens/json keyfiles
+  # NOTE: the FireCloudClient instance calling these methods must be initialized using a User object
+  # for authentication purposes, otherwise downstream calls will return 403
+  ##
+
+  # create a new instance of FireCloudClient, but use a pet service account keyfile
+  # will only work with users that have been registered in Terra
+  #
+  # * *params*
+  #   - +user+: (User) => User object from which access tokens are generated
+  #   - +project+: (String) => Default GCP Project to use (can be overridden by other parameters)
+  #
+  # * *return*
+  #   - +FireCloudClient+ instance, or nil if user has not registered with Terra
+  def self.new_with_pet_sa(user, project)
+    # create a temporary client in order to retrieve the user's pet service account keyfile
+    tmp_client = self.new(user, project)
+    if tmp_client.registered?
+      pet_service_account_json = tmp_client.get_pet_service_account_key(project)
+      self.new(user, project, pet_service_account_json)
+    else
+      nil
+    end
+  end
+
+  # issue an access_token for a user's pet service account in the requested project
+  #
+  # * *params*
+  #   - +project_name+ (String) => Name of a FireCloud billing project in which pet service account resides
+  #
+  # * *returns*
+  #   - +String+ pet service account OAuth2 access_token
+  def get_pet_service_account_token(project_name)
+    path = "https://sam.dsde-prod.broadinstitute.org/api/google/v1/user/petServiceAccount/#{project_name}/token"
```

Comment:
I recall that [a Bard problem we had last year](https://broadinstitute.slack.com/archives/CBEHTH601/p1600365389125200?thread_ts=1600365317.124600&cid=CBEHTH601) was caused by the (opaque-to-SCP) use of different SAMs for each environment -- i.e. development, staging, and production).

That makes me wary of hard-coding one SAM for all environments.  I'm not confident it'll be problematic, but if we encounter strange auth issues, it might be worth keeping this in mind.

Reply:
We would actually encounter auth issues if we used environment-specific domains.  We _always_ point at production for FireCloud API calls (e.g. api.firecloud.org), and if we pointed at a dev/test/alpha SAM instance, the two would not have the same list of users and associated data.  

FWIW, I talked with Kristian about this in the past - specifically should we point at dev in dev, alpha in staging, etc - and he emphatically said no.

Reply:
Helpful detail, thank you.  I'm no longer wary.  Given we use api.firecloud.org in all environments, it makes sense to use sam.dsde-prod there too.

Notably, that differs from how we use Bard.  [SCP uses environment-specific Bards](https://github.com/broadinstitute/single_cell_portal_core/search?q=bard_host_url) to keep production analytics accurate, and [Bard uses environment-specific SAMs](https://github.com/DataBiosphere/bard/search?q=sam).  

This means SCP implementers will get auth errors for Bard when signed in on development or staging, unless they've gone out of their way to [register in that environment's specific Terra](https://broadinstitute.slack.com/archives/CBEHTH601/p1600365389125200?thread_ts=1600365317.124600&cid=CBEHTH601).  But those are upstream analytics infrastructure issues.  They slightly overlap matters here, and have workarounds, and aren't problematic here.

Reply:
I agree that the current setup is not optimal given the mixed implementations for environments, but merging development and production analytics ones seems like an even larger issue as opposed to this one.